### PR TITLE
Create subway functions

### PR DESCRIPTION
**Pull request by kamatozzzik**

### CODE REVIEW

File: src/js/startApp.js [hunk 1]
```diff
@@ -1,17 +1,39 @@
-import { Station } from './Station';
-import { SubwayLine } from './SubwayLine';
-import { Subway } from './Subway';
-import { Route } from './Route';
+import { createSubway } from './subway func'
```
Comment (AndreiShenets) [inline src/js/startApp.js]: Do not use spaces in file names

File: src/js/startApp.js [hunk 2]
```diff
@@ -1,17 +1,39 @@
...
 textInput.addEventListener('change', function(e) {
-    const file = e.target.files[0];
+    let file = e.target.files[0];
     fileReader.readAsBinaryString(file);
 });
 
 fileReader.addEventListener('loadend', function() {
-	subwayData = this.result;
+    subwayData = this.result;
+    subwayData = parser.handleToArray(subwayData);
```
Comment (AndreiShenets) [inline src/js/startApp.js]: One method parse is expected. Should return complitely built model

File: src/js/subway func.js [hunk 3]
```diff
@@ -0,0 +1,48 @@
+import { Station } from './Station';
```
Comment (AndreiShenets) [inline src/js/subway func.js]: These functions should belong to Parser why did you extract them to separate file?

**Review by AndreiShenets — CHANGES_REQUESTED** [review]

File: src/js/startApp.js [hunk 4]
```diff
@@ -1,17 +1,39 @@
...
-import { Route } from './Route';
+import { createSubway } from './subway func'
+import { Parser } from './Parser'
 
 let fileReader = new FileReader();
+let parser = new Parser();
 let textInput = document.querySelector('#file-input');
+let currentSubway = null; 
 let subwayData = null;
+const lsKeys = {
```
Comment (AndreiShenets) [inline src/js/startApp.js]: it is worth to create separate constant for this. also it is worth to name it with full name localStorageKeys

**Review by AndreiShenets — CHANGES_REQUESTED** [review]

**Comment by kamatozzzik** [comment]
all fixed

File: package.json [hunk 5]
```diff
@@ -0,0 +1,26 @@
+{
+  "name": "subway",
+  "version": "1.0.0",
+  "description": "",
+  "main": "webpack.config.js",
```
Comment (AndreiShenets) [inline package.json:5]: Why this file is main? It is a start point of your application. Usually index.html or index.js

File: src/js/Parser.js [hunk 6]
```diff
@@ -0,0 +1,92 @@
+import { Station } from './Station';
+import { SubwayLine } from './SubwayLine';
+import { Subway } from './Subway';
+
+export function parseToSubway(data) {
```
Comment (AndreiShenets) [inline src/js/Parser.js]: why name is parseToSubway? parseSubwayData or just parse is ok. parseToSubway isn't like English phrase

File: src/js/Parser.js [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+    let currentData = data;
+
+    currentData = handleToArray(currentData);
+    currentData = handleToSubway(currentData);
+
```
Comment (AndreiShenets) [inline src/js/Parser.js:11]: unnecessary line

File: src/js/Parser.js [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+
+    return createSubway(currentData);
+} 
+
+function handleToArray(data) {
+    let array = data;
+    array = array.split('\n');
+    array = Array.from(array);
```
Comment (AndreiShenets) [inline src/js/Parser.js]: Why do you do this?

File: src/js/Parser.js [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+    let currentData = data;
+
+    currentData = handleToArray(currentData);
+    currentData = handleToSubway(currentData);
+
+
+    return createSubway(currentData);
+} 
+
+function handleToArray(data) {
```
Comment (AndreiShenets) [inline src/js/Parser.js]: handleToArray isn't like English phrase: convertToArray is more like what you want

File: src/js/Parser.js [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+    let array = data;
+    array = array.split('\n');
+    array = Array.from(array);
+
+    array = array.map(name => {
```
Comment (AndreiShenets) [inline src/js/Parser.js]: arrow functions support syntax like .map(name => name.trim())

File: src/js/Parser.js [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+        return name.trim();
+    });
+
+    return array;
+}
+
+function handleToSubway(data) {
```
Comment (AndreiShenets) [inline src/js/Parser.js]: handleToSubway isn't like English phrase. buildSubwayModel is more like what you want

File: src/js/Parser.js [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+    let stationNames = [], names = [], lineNames = [];
```
Comment (AndreiShenets) [inline src/js/Parser.js]: what does names array contain? it is hard to understand from the name

File: src/js/Parser.js [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+    let lineName = null;
+
+    data.forEach(name => {
+        if (!name || data.lastIndexOf(name) === data.length - 1) {
```
Comment (AndreiShenets) [inline src/js/Parser.js]: you introduce additonal algorithmic complexity using lastIndexOf. The first one you can use simple for and have index and determine index of current element without lastIndexOf. The second one forEach() has the second parameter index.
Also you call it twice!

File: src/js/Parser.js [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+            if (data.lastIndexOf(name) === data.length - 1) {
+                names.push(name);
+            }
+            lineName = names.shift();
+            stationNames.push(names);
+            lineNames.push(lineName);
+            names = [];
+        } else {
+        names.push(name);
```
Comment (AndreiShenets) [inline src/js/Parser.js]: Formatting issue

File: src/js/Parser.js [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+    array = array.map(name => {
+        return name.trim();
+    });
+
+    return array;
+}
+
+function handleToSubway(data) {
+    let stationNames = [], names = [], lineNames = [];
+    let lineName = null;
```
Comment (AndreiShenets) [inline src/js/Parser.js]: Why have you declared lineName here? I do not see usages apart from inside forEach

File: src/js/Parser.js [hunk 16]
```diff
@@ -0,0 +1,92 @@
...
+        } else {
+        names.push(name);
+        }
+    });
+
+    return { lineNames: lineNames, stationNames: stationNames };
+}
+
+function createSubway(data) {
+    let stationNames = data.stationNames.concat(),
```
Comment (AndreiShenets) [inline src/js/Parser.js]: Please do not use single let when you declare variables. Use separate let for each declaration

File: src/js/Parser.js [hunk 17]
```diff
@@ -0,0 +1,92 @@
...
+function handleToSubway(data) {
+    let stationNames = [], names = [], lineNames = [];
+    let lineName = null;
+
+    data.forEach(name => {
+        if (!name || data.lastIndexOf(name) === data.length - 1) {
+            if (data.lastIndexOf(name) === data.length - 1) {
+                names.push(name);
+            }
+            lineName = names.shift();
```
Comment (AndreiShenets) [inline src/js/Parser.js]: Why don't constract SubwayLine here?

File: src/js/Parser.js [hunk 18]
```diff
@@ -0,0 +1,92 @@
...
+        }
+    });
+
+    return { lineNames: lineNames, stationNames: stationNames };
+}
+
+function createSubway(data) {
+    let stationNames = data.stationNames.concat(),
+    stationList = [],
+    lines = data.lineNames.concat(),
```
Comment (AndreiShenets) [inline src/js/Parser.js]: Why do you do concat?

File: src/js/Parser.js [hunk 19]
```diff
@@ -0,0 +1,92 @@
...
+    allLines = [];
+
+    for (let i = 0; i < stationNames.length; i++) {
```
Comment (AndreiShenets) [inline src/js/Parser.js]: What do you do here?

File: src/js/Parser.js [hunk 20]
```diff
@@ -0,0 +1,92 @@
...
+
+        allLines[i] = createLine(lines[i], lineStations);
+        lineStations = [];
+    }
+
+    return new Subway(allLines);
+}
+
+
+function createLine(subwayName, stations) {
```
Comment (AndreiShenets) [inline src/js/Parser.js]: why do you use subwayName variable name if it is subwayLineName?

File: src/js/Parser.js [hunk 21]
```diff
@@ -0,0 +1,92 @@
...
+    let currentStations = stations.concat();
+    let line = new SubwayLine(subwayName,currentStations);
+    return line;
+}
+
+function createStation(name) {
```
Comment (AndreiShenets) [inline src/js/Parser.js]: why do you need this function? it is easier just to call new Station(name)

File: src/js/startApp.js [hunk 22]
```diff
@@ -1,17 +1,31 @@
...
-import { SubwayLine } from './SubwayLine';
-import { Subway } from './Subway';
-import { Route } from './Route';
+import { parseToSubway } from './Parser'
 
 let fileReader = new FileReader();
 let textInput = document.querySelector('#file-input');
-let subwayData = null;
+let currentSubway = null; 
+const localStorageKey = 'data';
```
Comment (AndreiShenets) [inline src/js/startApp.js]: localStorageKey of what? such constants should contain information about what they are key for

File: src/js/startApp.js [hunk 23]
```diff
@@ -1,17 +1,31 @@
...
+
 
 textInput.addEventListener('change', function(e) {
-    const file = e.target.files[0];
+    let file = e.target.files[0];
```
Comment (AndreiShenets) [inline src/js/startApp.js]: once again why let here if you do not change file reference?

File: src/js/startApp.js [hunk 24]
```diff
@@ -1,17 +1,31 @@
...
     fileReader.readAsBinaryString(file);
 });
 
 fileReader.addEventListener('loadend', function() {
-	subwayData = this.result;
+    currentSubway = parse.parseToSubway(this.result);
+
+    let storageObj = JSON.stringify(currentSubway);
```
Comment (AndreiShenets) [inline src/js/startApp.js]: why let here if you do not change object reference?

File: src/js/startApp.js [hunk 25]
```diff
@@ -1,17 +1,31 @@
...
+    localStorage.setItem(localStorageKey, storageObj);
+
```
Comment (AndreiShenets) [inline src/js/startApp.js]: unnecessary line

File: src/js/startApp.js [hunk 26]
```diff
@@ -1,17 +1,31 @@
...
 });
+
+window.addEventListener('load', () => {
+    let localStorageData = localStorage.getItem(localStorageKey);
+
+    if (localStorageData) {
+        currentSubway = JSON.parse(localStorageData);
+    }
+});
+
```
Comment (AndreiShenets) [inline src/js/startApp.js]: two unnecessary lines. only one trailing line is acceptable

File: src/js/Parser.js [hunk 27]
```diff
@@ -0,0 +1,92 @@
...
+        allLines[i] = createLine(lines[i], lineStations);
+        lineStations = [];
+    }
+
+    return new Subway(allLines);
+}
+
+
+function createLine(subwayName, stations) {
+    let currentStations = stations.concat();
```
Comment (AndreiShenets) [inline src/js/Parser.js]: Why do you do concat?

File: src/js/Parser.js (hunk 20 above)
Comment (AndreiShenets) [inline src/js/Parser.js]: Why do you need this function?

File: src/js/Parser.js [hunk 28]
```diff
@@ -0,0 +1,92 @@
...
+            if (!stationList.includes(stationNames[i][j]))
+                stationList.push(stationNames[i][j]);
+        }
+    }
+
+    stationList = stationList.map(name => {
+        return createStation(name);
+    });
+
+    for (let i = 0; i < lines.length; i++) {
```
Comment (AndreiShenets) [inline src/js/Parser.js]: Please explaing what is going on in this function...

**Review by AndreiShenets — CHANGES_REQUESTED** [review]

File: src/js/Parser.js [hunk 29]
```diff
@@ -0,0 +1,74 @@
...
+	for(let i = 0; i < data.length; i++) {
+		if (!data[i] || i === data.length - 1) {
+			if (i === data.length - 1) {
+				allNames.push(data[i]);
+			}
+			lineName = allNames.shift();
+			stationNames.push(allNames);
+			lineNames.push(lineName);
+
+			names = [];
```
Comment (AndreiShenets) [inline src/js/Parser.js]: Tabs, forSPACE(, names not exist

File: src/js/Parser.js [hunk 30]
```diff
@@ -0,0 +1,53 @@
+import { Station } from './Station';
+import { SubwayLine } from './SubwayLine';
+import { Subway } from './Subway';
+
+export function parse(data) {
+    let currentData = data;
```
Comment (AndreiShenets) [inline src/js/Parser.js]: you do not need to store data to local variable you can just pass it to you function. after thatyou even do not need convertToSubway function you can move it's content to parse function

File: src/js/Parser.js [hunk 31]
```diff
@@ -0,0 +1,53 @@
...
+    return convertToSubway(currentData);
+} 
+
+function convertToSubway(data) {
+    let allNames = [];
+    let stationNames = [];
+    let subwayLines = [];
+    let result = {};
+    data = data.split('\n').map(name => name.trim());
```
Comment (AndreiShenets) [inline src/js/Parser.js:10]: I would add one empty line above and place each chain method to separate line

File: src/js/Parser.js [hunk 32]
```diff
@@ -0,0 +1,53 @@
...
+
+    for (let i = 0; i < data.length; i++) {
+        if (!data[i] || i === data.length - 1) {
+            if (i === data.length - 1) {
+                allNames.push(data[i]);
+            }
+            let lineName = allNames.shift();
+            stationNames = stationNames.concat(allNames);
```
Comment (AndreiShenets) [inline src/js/Parser.js]: Why do you do concat? it seems that you do not need stationNames variable at all. Let's discuss the logic again

**Review by AndreiShenets — CHANGES_REQUESTED** [review]

File: src/js/Parser.js [hunk 33]
```diff
@@ -0,0 +1,43 @@
...
+import { SubwayLine } from './SubwayLine';
+import { Subway } from './Subway';
+
+export function parse(data) {
+    let allNames = [];
+    let stationNames = [];
+    let subwayLines = [];
+    let result = {};
+    data = data
+            .split('\n')
```
Comment (AndreiShenets) [inline src/js/Parser.js]: One tab shift

**Review by AndreiShenets — CHANGES_REQUESTED** [review]

File: src/js/Parser.js [hunk 34]
```diff
@@ -0,0 +1,37 @@
...
+
+    for (let i = 0; i < data.length; i++) {
+        if (!data[i] || i === data.length - 1) {
+            if (i === data.length - 1) {
+                allNames.push(data[i]);
+            }
+            let lineName = allNames.shift();
+
+            allNames.forEach(name => {
+                stationNames[name] = new Station(name);
```
Comment (AndreiShenets) [inline src/js/Parser.js]: If name already exists you recreate object...

File: src/js/Parser.js [hunk 35]
```diff
@@ -0,0 +1,37 @@
...
+            });
+            result[lineName] = allNames;
```
Comment (AndreiShenets) [inline src/js/Parser.js]: The issue above not allow you set stations here

File: src/js/Parser.js [hunk 36]
```diff
@@ -0,0 +1,37 @@
...
+            allNames = [];
+        } else {
+            allNames.push(data[i]);
+        }
+    }
+
+    for (let key in result) {
+        let stations = [];
+
+        stations = result[key].map(name => stationNames[name]);
```
Comment (AndreiShenets) [inline src/js/Parser.js]: The issue above force you to do this mapping

**Review by AndreiShenets — CHANGES_REQUESTED** [review]

**Review by AndreiShenets — APPROVED** [review]

**Pull request merged**